### ISSUE TITLE
Change font in print

### ISSUE THREAD

**Issue by bbahes** [joncampbell123/dosbox-x#2305]
HI!
I'm using Dosbox-x on WIndows 10 for old DOS application to print invoices. 
So far I was able to setup it to print to either pdf or printer connected to host.
However, I'm unable to change font using fontpath in dosbox-x.conf.txt. The problem with print is that it does not have Croatian characters, so I wanted to change font. 
Is this something related to how application sends to printer, which I'm not sure how it does. I believe it creates PRN file before.

Can anyone help with this?
Thanks in advance for any help!

Regards.
Boris.

**Comment by ThePillenwerfer** [comment]
I don't know if you've made an typing error above but the file you need to change is `dosbox-x.conf` not `dosbox-x,conf.txt`

**Comment by bbahes** [comment]
Yes, it's typo. "dosbox-x -printconf" does not output anything. I was able to change ttf.font value to font in FONTS sub directory, however not the print output. Could it be that it can't be changed? Reading https://github.com/Wengier/dosbox-x-wiki/wiki/Guide%3ASetting-up-printing-in-DOSBox%E2%80%90X and I see:

Limitations:

Windows printer output in black & white only

Not all ESC/P commands are supported, **like custom fonts**

Due to over-exact graphics emulation 360dpi printing from Windows (guest) graphics rastering might not look as expected

Country code setup and other configuration switches and buttons found on a real printer are not available, this may be overcome by sending special ESC/P commands to the printer before printing

**Comment by rderooy** [comment]
Custom fonts in this context refers to the Epson ESC/P language supporting custom fonts. I think this refers to optional font cartridges that some printers supported back in the day. It has no relation to TrueType fonts.

What DOS application are you using for printing, and which printer did you configure? Does the application print in text mode or in graphics mode? Text mode means it simply outputs text with some control codes, where graphics mode is more common for later (largly Windows) applications to support more fancy fonts.

Assuming the application prints in text mode, there may also be a possible codepage translation issue. If you print to a text file, does it contain right Croatian characters?

**Comment by ThePillenwerfer** [comment]
You can certainly change the font.  The limitation in the wiki means that you can't embed an escape code in a document to change the font in part of it whilst it's actually printing.

Copying your chosen font to the `FONTS` sub-directory and renaming it `courier.ttf` should definitely work with no change to the `.conf` file.

I can't think why `dosbox-x -printconf` doesn't say anything though. 

Can you post the `[printer]` section of your `.conf` file?

**Comment by bbahes** [comment]
I had to reinstall entire Dosbox-X because of dosprint.com messed up config.

printer     = true
dpi         = 360
width       = 85
height      = 110
printoutput = printer
multipage   = false
device      = -
docpath     = .
fontpath    = FONTS\courier.ttf
openwith    = 
openerror   = 
shellhide   = false
timeout     = 0

FONTS directory:
![image](https://user-images.githubusercontent.com/19546260/108920983-f1fa7780-7635-11eb-982a-e2ffc0e3c264.png)

Croatian letters in edit.com
![image](https://user-images.githubusercontent.com/19546260/108921012-fde63980-7635-11eb-84dc-292364e13033.png)

Printing from edit.com produces Croatian letters
![image](https://user-images.githubusercontent.com/19546260/108921179-41d93e80-7636-11eb-98c8-bfab1bde8772.png)


but from application they are missing:

![image](https://user-images.githubusercontent.com/19546260/108921314-6df4bf80-7636-11eb-848b-f41a66070cb0.png)

**Comment by bbahes** [comment]
> Custom fonts in this context refers to the Epson ESC/P language supporting custom fonts. I think this refers to optional font cartridges that some printers supported back in the day. It has no relation to TrueType fonts.
> 
> What DOS application are you using for printing, and which printer did you configure? Does the application print in text mode or in graphics mode? Text mode means it simply outputs text with some control codes, where graphics mode is more common for later (largly Windows) applications to support more fancy fonts.
> 
> Assuming the application prints in text mode, there may also be a possible codepage translation issue. If you print to a text file, does it contain right Croatian characters?

This is old accounting application written in Croatia by company that does not support it anymore. I'm not sure is it text or graphics mode but I'm aiming for text mode since they used it in Windows 98 environment and printed to Epson LQ300 printer.

**Comment by bbahes** [comment]
Hm....I got it to change font, however, still no Croatian letters:

![image](https://user-images.githubusercontent.com/19546260/108923600-671b7c00-7639-11eb-983f-521884e52024.png)

**Comment by bbahes** [comment]
And finally...got it by downloading weird font from http://www.dusko-lolic.from.hr/unihr/ ISO-8859-2, WIN-1250, CROSCII...maybe you can clarify bit more?

![image](https://user-images.githubusercontent.com/19546260/108924409-dfcf0800-763a-11eb-9310-5cc72f6d6a9b.png)

**Comment by ThePillenwerfer** [comment]
Glad you got it working.

The only thing I'd say about your `[printer]` section is that `fontpath` is the path to where the fonts are, not the font itself, ie `FONTS` rather than `FONTS\courier.ttf`.  I also see that you are using settings for American _letter_ paper rather than A4.  If you are using A4 set it:—
```
width       = 82
height      = 117
```
DOSBox-X's printing is still a bit rough 'round the edges but hopefully will improve with time.

**Comment by Wengier** [comment]
Yes, ```fontpath```  is the path to the fonts, not the font itself. I think this is more a documentation issue - the comments/documentation can be improved to make this more clear to the user.

**Comment by rderooy** [comment]
I'm a little confused, is this a DOS application or a Windows 3.x or Win9x application?

Looking at the output, this seems like a codepage issue as the Croatian characters are mapped to other ASCII characters.

The download you pointed to (unihr7.zip) has a TrueType font. Did you put that into the DOSBox-X fonts directory and rename it to COURIER.TTF or did you use it inside DOSBox-X with Windows 3/x/Win9x?

**Comment by bbahes** [comment]
This is DOS application. Yes I've downloaded this font and renamed it to courier.ttf in FONTS directory.
I've downloaded 7-bit CROSCII font and after that I got Croatian letters.

**Comment by rderooy** [comment]
With that special font, if you now print from EDIT, do you still get the Croatian characters?

I suspect not. I think the DOS application is expecting a printer with a strange codepage (not 850 or 437), and that 7-bit font just happens to have the Croatian characters at the right location.

I had asked this before, but did not see an answer. What printer is configured in the DOS application?

**Comment by bbahes** [comment]
No Croatian letters from edit.com now that I changed font.
From what I see (by going through DBF file that seems to hold printer config) it uses .PRN file that emulates EPSON.
[EPS852.PRN.txt](https://github.com/joncampbell123/dosbox-x/files/6054913/EPS852.PRN.txt)

**Comment by rderooy** [comment]
I think the 852 in the filename refers to codepage 852. What keyboard/codepage have you configured in DOSBox-X?

Also, the SDL2 build has the best support for non-US keyboard layouts. With the SDL1 build you may encounter issues with some keys.

Back in the DOS days, Croatia was still part of Yugoslavia and this is reflected here.

```
Z:\>keyb yu
Keyboard layout yu loaded for codepage 852
```

If you use a different keyboard layout, you can specify the codepage as such:

```
Z:\>keyb de 852
Keyboard layout de loaded for codepage 852
```
Which will give you a German layout but with codepage 852.

See also:
https://www.dosbox.com/wiki/KEYB

**Comment by bbahes** [comment]
You are correct.
Currently keyb.com reports:
![image](https://user-images.githubusercontent.com/19546260/109394722-ba871600-7928-11eb-82ce-7999098a37bb.png)

Note foggs.com utility. This utility was put there to emulate Croatian letters in application layout:
![image](https://user-images.githubusercontent.com/19546260/109394772-f8843a00-7928-11eb-9ba9-5b90fb1733c1.png)

If not:
![image](https://user-images.githubusercontent.com/19546260/109394792-1d78ad00-7929-11eb-9690-5bb9a8e2e2ca.png)

**Comment by Wengier** [comment]
@bbahes If you use the TrueType font (TTF) output which is recommended for text-mode DOS application usage, and with the code page 852 you probably won't need any external utility (like foggs.com) in order to display Croatian letters. For this to work you can just set ```output=ttf``` and maybe ```chcp 852```.

**Comment by bbahes** [comment]
I tried that, but could not get it to work. 
When I set "output=ttf" I get:

![image](https://user-images.githubusercontent.com/19546260/109396534-62eda800-7932-11eb-94f6-0c41bab5f919.png)

**Comment by Wengier** [comment]
@bbahes I see that you must be using DOSBox-X 0.83.10, which did not have CHCP command so it won't work. The CHCP command was added for DOSBox-X 0.83.11 which should be released very shortly. Since you use Windows 10, can you try the following DOSBox-X binary package which provides CHCP command:

* [dosbox-x-update.zip](http://cndos.fam.cx/dosware/dosbox-x-update.zip)

With this try the TTF output and the command ```CHCP 852``` and see how it works.

**Comment by bbahes** [comment]
I've replaced files in DOSBox-X directory, however, when I try to launch it, it just opens main window and closes:
![image](https://user-images.githubusercontent.com/19546260/109397142-823a0480-7935-11eb-9f13-81a389d5a0d5.png)

**Comment by Wengier** [comment]
The DOSBox-X build included in the package is a 32-bit one. Have you tried to extract the files to a separate directory instead? Alternatively, you can just wait for the official 0.83.11 release which should be ready by the end of this month.

**Comment by bbahes** [comment]
I've extracted to separate directory now and same. It opens and closes.
So it's today or tomorrow? :)

**Comment by Wengier** [comment]
It may be related to some setting. For now can you try the following from the command-line to launch DOSBox-X:

```
dosbox-x -defaultconf -set output=ttf -c "chcp 852"
```

Yes, the new version is expected to come on the last day of the month.

**Comment by bbahes** [comment]
Upgraded to 0.83.11 result is:

![image](https://user-images.githubusercontent.com/19546260/109541751-703a9c00-7ac4-11eb-93ef-f1e91102791f.png)

![image](https://user-images.githubusercontent.com/19546260/109541829-86485c80-7ac4-11eb-9a8d-965e6df492b2.png)

**Comment by Wengier** [comment]
@bbahes Interesting. Could you upload the Croatian language program so that I can try it myself? Thanks.

**Comment by bbahes** [comment]
No problem. Please contact me on e-mail.

**Comment by Wengier** [comment]
@bbahes My email address is [this](wengierwu@yahoo.com). You can send it here.
